### PR TITLE
feat: code analysis

### PR DESCRIPTION
**Pull request by echosoar**

### CODE REVIEW

File: src/comprehension/analysis.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import { geNodeInfo } from './node';
+
+interface IAnalysisOptions {
+  program?: Program;
+  compilerHost?: CompilerHost;
+  projectDir?: string;
+  mwccConfig?: MwccConfig;
+  decoratorLowerCase?: boolean;
+}
+export class Analysis {
```
Comment (legendecas) [inline src/comprehension/analysis.ts:16]:
```suggestion
export class Analyzer {
```

File: src/comprehension/analysis.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  private options: IAnalysisOptions;
+  private checker: ts.TypeChecker;
+
+  constructor(options: IAnalysisOptions) {
+    this.options = options || {};
+    this.program = this.initProgram(options);
+    this.checker = this.program.getTypeChecker();
+  }
+
+  public analysis() {
```
Comment (legendecas) [inline src/comprehension/analysis.ts:29]:
```suggestion
  public analyze() {
```

File: src/comprehension/analysis.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  program?: Program;
+  compilerHost?: CompilerHost;
+  projectDir?: string;
+  mwccConfig?: MwccConfig;
+  decoratorLowerCase?: boolean;
+}
+export class Analysis {
+
+  private program: Program;
+  private analysisResult = {};
```
Comment (legendecas) [inline src/comprehension/analysis.ts:19]: Missing type definition

File: src/comprehension/node.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+const getType = (symbol: ts.Symbol, checker) => {
+  const checkerType: ts.Type = checker.getTypeOfSymbolAtLocation(
+    symbol,
+    symbol.valueDeclaration,
+  );
+  return {
+    type: checker.typeToString(checkerType).toLowerCase(),
+    checkerType,
+  };
+}
```
Comment (legendecas) [inline src/comprehension/node.ts:56]: Missing EOF

File: src/comprehension/params.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      const param: any = {};
+      if (symbol.members) {
+        symbol.members.forEach((value, key) => {
+          param[key as string] = getParamBySymbol(value);
+        });
+      }
+      return param;
+  }
+  return text;
+}
```
Comment (legendecas) [inline src/comprehension/params.ts:76]: Ditto

File: src/util.ts [hunk 6]
```diff
@@ -82,3 +82,24 @@ export function safeJsonParse<T = any>(str: string): T | undefined {
...
+    const codeArr: string[] = code.substr(0, pos).split('\n');
+    const ln = codeArr.length - 1;
+    return {
+      ln,
+      col: codeArr[ln]?.length || 0,
+      index: pos,
+    };
+  } catch {  }
+  return { ln: 0, col: 0, index: 0 };
+}
```
Comment (legendecas) [inline src/util.ts:105]: Ditto

Comment (echosoar) [reply src/util.ts]: Fixed

File: src/comprehension/params.ts (hunk 5 above)
Comment (echosoar) [reply src/comprehension/params.ts]: Fixed

File: src/comprehension/node.ts (hunk 4 above)
Comment (echosoar) [reply src/comprehension/node.ts]: Fixed

File: src/comprehension/analysis.ts (hunk 3 above)
Comment (echosoar) [reply src/comprehension/analysis.ts]: Fixed

File: src/comprehension/analysis.ts (hunk 2 above)
Comment (echosoar) [reply src/comprehension/analysis.ts]: Fixed

File: src/comprehension/analysis.ts (hunk 1 above)
Comment (echosoar) [reply src/comprehension/analysis.ts]: Fixed

**Pull request merged**